test/test_device_batch_memcpy.cu

```diff
@@ -0,0 +1,548 @@
+  std::vector<BufferSizeT> h_buffer_sizes(num_buffers);
+  std::vector<ByteOffsetT> h_buffer_src_offsets(num_buffers);
+  std::vector<ByteOffsetT> h_buffer_dst_offsets(num_buffers);
+
+  // Device-side resources
+  void *d_in                        = nullptr;
+  void *d_out                       = nullptr;
+  ByteOffsetT *d_buffer_src_offsets = nullptr;
+  ByteOffsetT *d_buffer_dst_offsets = nullptr;
+  BufferSizeT *d_buffer_sizes       = nullptr;
+  void *d_temp_storage              = nullptr;
+  size_t temp_storage_bytes         = 0;
+
+  cudaEvent_t events[2];
+  cudaEventCreate(&events[0]);
+  cudaEventCreate(&events[1]);
+
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
+
+  // Generate the buffer sizes
+  GenerateRandomData(h_buffer_sizes.data(), h_buffer_sizes.size(), min_buffer_size, max_buffer_size);
+
+  // Make sure buffer sizes are a multiple of the most granular unit (one AtomicT) being copied
+  // (round down)
+  for (BufferOffsetT i = 0; i < num_buffers; i++)
+  {
+    h_buffer_sizes[i] = (h_buffer_sizes[i] / sizeof(AtomicT)) * sizeof(AtomicT);
+  }
+
+  // Compute the total bytes to be copied
+  ByteOffsetT num_total_bytes = 0;
+  for (BufferOffsetT i = 0; i < num_buffers; i++)
+  {
+    if (input_gen == TestDataGen::CONSECUTIVE)
+    {
+      h_buffer_src_offsets[i] = num_total_bytes;
+    }
+    if (output_gen == TestDataGen::CONSECUTIVE)
+    {
+      h_buffer_dst_offsets[i] = num_total_bytes;
+    }
+    num_total_bytes += h_buffer_sizes[i];
+  }
+
+  // The input buffer's total size
+  data_size = num_total_bytes;
+
+  // Shuffle input buffer source-offsets
+  std::uint_fast32_t shuffle_seed = 320981U;
+  if (input_gen == TestDataGen::RANDOM)
+  {
+    h_buffer_src_offsets = GetShuffledBufferOffsets<BufferOffsetT, ByteOffsetT>(h_buffer_sizes,
+                                                                                shuffle_seed);
+    shuffle_seed += 42;
+  }
+
+  // Shuffle input buffer source-offsets
+  if (output_gen == TestDataGen::RANDOM)
+  {
+    h_buffer_dst_offsets = GetShuffledBufferOffsets<BufferOffsetT, ByteOffsetT>(h_buffer_sizes,
+                                                                                shuffle_seed);
+  }
+
+  // Allocate device memory
+  CubDebugExit(cudaMalloc(&d_in, data_size));
+  CubDebugExit(cudaMalloc(&d_out, num_total_bytes));
+  CubDebugExit(cudaMalloc(&d_buffer_src_offsets, num_buffers * sizeof(d_buffer_src_offsets[0])));
+  CubDebugExit(cudaMalloc(&d_buffer_dst_offsets, num_buffers * sizeof(d_buffer_dst_offsets[0])));
+  CubDebugExit(cudaMalloc(&d_buffer_sizes, num_buffers * sizeof(d_buffer_sizes[0])));
+
+  // Populate the data source with random data
+  using RandomInitAliasT      = uint16_t;
+  uint32_t num_aliased_factor = sizeof(RandomInitAliasT) / sizeof(uint8_t);
+  uint32_t num_aliased_units  = CUB_QUOTIENT_CEILING(data_size, num_aliased_factor);
+  uint8_t *h_in               = new uint8_t[num_aliased_units * num_aliased_factor];
```

Comment:
If I understand correctly, you allocate memory but not free it. Let's use `unique_ptr<uint8_t[]>` here.

Reply:
Thanks, good catch. Made these `unique_ptr`.